src/librustc_middle/traits/mod.rs

```diff
@@ -80,8 +81,33 @@ pub enum Reveal {
 }
 
 /// The reason why we incurred this obligation; used for error reporting.
+///
+/// As the happy path does not care about this struct, we store it on the heap
+/// to improve the performance of the happy path.
+///
+/// We do not want to intern this as there are a lot of obligation causes which
+/// only live for a short period of time.
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
 pub struct ObligationCause<'tcx> {
+    data: Lrc<ObligationCauseData<'tcx>>,
```

Comment:
Did you consider `Cow<'tcx, ObligationCauseData<'tcx>>`, with the borrowed variant being used for `ObligationCause::dummy`? This would require two pointers instead of one, but doesn't need synchronization when using the parallel compiler.

Reply:
You could also use a custom `enum` with a `Dummy` variant whose `Deref` impl points to a `static` with the dummy `ObligationCauseData`. I feel like you maybe considered this but ruled it out? Presumably it breaks `make_mut`?

Reply:
I did not consider that :thinking: let me try implementing this using an `enum`

Reply:
seems nice :+1: